### PR TITLE
Add value checking of data, and add check for negative etenergies

### PR DESCRIPTION
**Pull request by langner**
This implements a pattern for the data checks described in #237, and adds a check for negative values in `etenergies`. Initially I added `check_values` to the parser class, but I think checks like negative `etenergies` are not specific to any one parser and so belong in the data class. And, they are probably best done after parsing is done, since things may change during parsing. That being said, there may be things in #237 that clearly belong on the parser side (we have some already). Let's deal with that on a case by case basis.

If this looks OK I'll add more checks.

### CODE REVIEW

File: test/parser/testdata.py [hunk 1]
```diff
@@ -23,11 +29,18 @@ class ccDataTest(unittest.TestCase):
...
 
     def _set_attributes(self, names, val):
         for n in names:
             setattr(self.data, n, val)
 
+    def test_valuecheck_negative_etenergies(self):
+        self.data.etenergies = [1, -1]
+        self.data.check_values(logger=self.mock_logger)
+        self.mock_logger.error.assert_called_once_with(
+            AnyStringWith("At least one excitation energy is negative"))
```
Comment (berquist) [inline test/parser/testdata.py]: So this is similar to possible behavior using https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaisesRegex, where the check is that part of the message contains `At least one excitation energy is negative`?

File: test/parser/testdata.py [hunk 2]
```diff
@@ -1,19 +1,25 @@
...
 
+import mock
 import unittest
 
 import numpy
 
 import cclib
 
 
+class AnyStringWith(str):
```
Comment (berquist) [inline test/parser/testdata.py]: I understand exactly what this does from reading `__eq__`, but not the class name. Can you change it to something containing `Contains` or similar?

File: cclib/parser/data.py [hunk 3]
```diff
@@ -290,6 +291,15 @@ def typecheck(self):
...
                 raise TypeError("attribute %s is %s instead of %s and could not be converted" % args)
 
+    def check_values(self, logger=logging):
+        """Perform custom checks on the values of attributes."""
+        if hasattr(self, "etenergies") and any(e < 0 for e in self.etenergies):
+            negative_values = [e for e in self.etenergies if e < 0]
+            msg = ("At least one excitation energy is negative. "
+                   "\nNegative values: %s\nFull etenergies: %s"
+                   % (negative_values, self.etenergies))
+            logger.error(msg)
```
Comment (berquist) [inline cclib/parser/data.py]: Using `ERROR` doesn't halt parsing, right?

File: cclib/parser/logfileparser.py [hunk 4]
```diff
@@ -434,19 +437,28 @@ def new_internal_job(self):
...
+        
+        Parameters
+        ----------
+        name: str
+            The name of the attribute.
+        value: str
+            The value for the attribute.
+        check_change: bool
+            By default we want to check that the value does not change
+            if the attribute already exists.
```
Comment (berquist) [inline cclib/parser/logfileparser.py]: Are we going to stick to NumPy-style docstrings? If we pick something, it should go in the developer docs.

File: test/parser/testdata.py [hunk 5]
```diff
@@ -1,19 +1,25 @@
...
 #
-# Copyright (c) 2017, the cclib development team
+# Copyright (c) 2019, the cclib development team
 #
 # This file is part of cclib (http://cclib.github.io) and is distributed under
 # the terms of the BSD 3-Clause License.
 
 """Unit tests for parser data module."""
 
+import mock
```
[the remark this answers is not in the document]
Comment (langner) [reply test/parser/testdata.py]: I'd prefer to use six for this kind of thing - done in 47454d3 for both modules. We could consolidate that in one place, but I don't think it's worth the effort for a few lines.

File: test/parser/testdata.py (hunk 1 above)
Comment (langner) [reply test/parser/testdata.py]: Right. I didn't want to put in the whole string verbatim, and I don't think there's anything built in like this in mock.

File: test/parser/testdata.py (hunk 2 above)
Comment (langner) [reply test/parser/testdata.py]: Changed to `StringContains`, hopefully that's good.

File: cclib/parser/data.py (hunk 3 above)
Comment (langner) [reply cclib/parser/data.py]: Not by default, no.

File: cclib/parser/logfileparser.py (hunk 4 above)
Comment (langner) [reply cclib/parser/logfileparser.py]: Added to #555. I'm not convinced by any particular format, but I'm OK with NumPy.

Comment (berquist) [reply cclib/parser/logfileparser.py]: I've written each of those forms, I have no preference.

**Pull request merged**